### PR TITLE
Sort datasets by parent and children, harvest parents first

### PR DESCRIPTION
**Pull request by avdata99**
Related to [Catalog#41](https://github.com/GSA/catalog.data.gov/issues/41) and [Catalog#70](https://github.com/GSA/catalog.data.gov/issues/70) In the previous version of catalog, while we harvest a DCAT-US source with collections (parent and children datasets), we divide the harvest process into two parts: First parent and the children.
We need to override the harvest base extension to fit this behaviour. We don't want to continue doing this.

This PR:
 - [divide DCAT-US](https://github.com/GSA/ckanext-datajson/blob/divide_dcat-us_harvester/ckanext/datajson/harvester_base.py#L4-L7) for different CKAn versions.
 - [Divide tests](https://github.com/GSA/ckanext-datajson/tree/divide_dcat-us_harvester/ckanext/datajson/tests): One for all harvesters, one for CKAN 2.8 and one for CKAN 2.3.
 - Hide the `datajson_collection` and `parent_run` feature for CKAN 2.8
 - Change the order of the HarvestObjest to ensure we will harvest parents first
 - Ensure we detect the `pakage_id` of the parent based on the `isPartOf` previous value.
 - Add a test case which covers this new feature

### CODE REVIEW

File: ckanext/datajson/harvester_base.py [hunk 1]
```diff
@@ -422,27 +387,28 @@ def import_stage(self, harvest_object):
                 is_collection = True
             if extra.key == 'collection_pkg_id' and extra.value:
                 parent_pkg_id = extra.value
+                if parent_pkg_id.startswith('IPO:'):
+                    # it's an IsPartOf ("identifier" at the external source)
+                    log.info('IPO found {}'.format(parent_pkg_id))
+
+                    #  check if parent is already harvested
+                    parent_identifier = parent_pkg_id.replace('IPO:', '') 
+                    expected_value = '"identifier": "{}"'.format(parent_pkg_id)
```
Comment (avdata99) [inline ckanext/datajson/harvester_base.py:396]:
```suggestion
                    expected_value = '"identifier": "{}"'.format(parent_identifier)
```

File: ckanext/datajson/tests/mock_datajson_source.py [hunk 2]
```diff
@@ -28,6 +28,12 @@ def do_GET(self):
         elif self.path == '/usda':
             self.sample_datajson_file = 'usda.gov.data.json'
             self.test_name = 'usda'
+        elif self.path == '/collections':
+            self.sample_datajson_file = 'collections.data.json'
+            self.test_name = 'collections'
+        elif self.path == '/collections2':
+            self.sample_datajson_file = 'collections2.data.json'
+            self.test_name = 'collections2'
```
Comment (adborden) [inline ckanext/datajson/tests/mock_datajson_source.py:36]: Can we give these better names that describe the scenario they represent? e.g. what's the difference between collections and collections2?

File: ckanext/datajson/harvester_base.py [hunk 3]
```diff
@@ -401,6 +366,59 @@ def _size_check(self, key, value):
...
             ))
+    
+    def get_harvest_source_id(self, package_id):
+        harvest_object = model.Session.query(HarvestObject) \
+            .filter(HarvestObject.package_id == package_id) \
+            .filter(HarvestObject.current==True).first()
+
+        return harvest_object.source.id if harvest_object else None
+
+    def is_part_of_to_package_id(self, ipo, harvest_object):
```
Comment (adborden) [inline ckanext/datajson/harvester_base.py:377]: Would you please add unit tests for these methods?

File: ckanext/datajson/tests/test_datajsonharvester.py [hunk 4]
```diff
@@ -42,94 +35,254 @@ def setup(cls):
...
+        for dataset in datasets:
+            assert dataset.title in titles
+            
+            is_parent = dataset.extras.get('collection_metadata', 'false').lower() == 'true'
+            is_child = dataset.extras.get('collection_package_id', None) is not None
+
+            log.info('Harvested dataset {} {} {}'.format(dataset.title, is_parent, is_child))
+
+            if dataset.title in ['Employee Relations Roundtables']:
+                assert_equal(is_parent, True)
```
Comment (adborden) [inline ckanext/datajson/tests/test_datajsonharvester.py:169]:
```suggestion
                assert_equal(is_parent, True)
                assert_equal(is_child, False)
```

File: ckanext/datajson/tests/test_datajsonharvester.py [hunk 5]
```diff
@@ -42,94 +35,254 @@ def setup(cls):
...
+                parent_counter += 1
+            else:
+                assert_equal(is_child, True)
```
Comment (adborden) [inline ckanext/datajson/tests/test_datajsonharvester.py:172]:
```suggestion
                assert_equal(is_child, True)
                assert_equal(is_parent, False)
```

**Comment by adborden** [comment]
... and it would be a good idea to add a test case for concurrency. Assert that if the child gets processed before a parent, it will get re-processed successfully.

File: ckanext/datajson/harvester_base.py [hunk 6]
```diff
@@ -401,6 +366,59 @@ def _size_check(self, key, value):
...
+                else:
+                    log.info('{} not found at {} for {}'.format(harvest_source.id, dataset_harvest_source_id, ipo))
+
+            msg = 'Unable to identify parent for: "{}" ({})'.format(ipo, results['count'])
+            harvest_object_error = HarvestObjectError(message=msg, object=harvest_object)
+            harvest_object_error.save()
+            log.error(msg)
+            return False, None
+        else:
+            return True, results['results'][0]
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/datajson/harvester_base.py]: fixed

File: ckanext/datajson/tests/mock_datajson_source.py (hunk 2 above)
Comment (avdata99) [reply ckanext/datajson/tests/mock_datajson_source.py]: Updated

File: ckanext/datajson/harvester_base.py [hunk 7]
```diff
@@ -195,9 +195,7 @@ def gather_stage(self, harvest_job):
         # first, mark the status in harvest_source config, which
         # triggers a children harvest_job after parents job is finished.
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/datajson/harvester_base.py]: Cleaned

File: ckanext/datajson/harvester_base.py [hunk 8]
```diff
@@ -422,27 +440,21 @@ def import_stage(self, harvest_object):
...
             if extra.key == 'collection_pkg_id' and extra.value:
                 parent_pkg_id = extra.value
+                if parent_pkg_id.startswith('IPO:'):
+                    # it's an IsPartOf ("identifier" at the external source)
+                    log.info('IPO found {}'.format(parent_pkg_id))
+
+                    #  check if parent is already harvested
+                    parent_identifier = parent_pkg_id.replace('IPO:', '') 
+                    parent_is_harvested, parent = self.is_part_of_to_package_id(parent_identifier, harvest_object)
+                    if not parent_is_harvested:
```
[the remark this answers is not in the document]
Comment (avdata99) [reply ckanext/datajson/harvester_base.py]: [This commit](https://github.com/GSA/ckanext-datajson/pull/66/commits/2fd993323b22800e82595a456171460a4a801c00) back this function to just one result, no more tuple

File: ckanext/datajson/harvester_base.py (hunk 3 above)
Comment (avdata99) [reply ckanext/datajson/harvester_base.py:374]: [This commit](https://github.com/GSA/ckanext-datajson/pull/66/commits/2fd993323b22800e82595a456171460a4a801c00) add some unit tests. I will add more

File: ckanext/datajson/harvester_base.py [hunk 9]
```diff
@@ -401,7 +423,64 @@ def _size_check(self, key, value):
...
+        
+        if results['count'] == 0:
+            msg = 'Parent identifier not found: "{}"'.format(ipo)
+            try:
+                harvest_object_error = HarvestObjectError(message=msg, object=harvest_object)
+                harvest_object_error.save()
+            except:
+                pass
+            log.error(msg)
+            raise ParentNotHarvestedException('Unable to find parent dataset. Raising error to allow re-run later')
```
Comment (adborden) [inline ckanext/datajson/harvester_base.py:452]: :clap:

File: ckanext/datajson/harvester_base.py [hunk 10]
```diff
@@ -260,20 +267,22 @@ def gather_stage(self, harvest_job):
...
-                else:
-                    # which is 'children_run'.
-                    # error out since parents got issues.
-                    self._save_gather_error(
-                        "Record with identifier '%s': isPartOf '%s' points to \
-                        an erroneous record." % (dataset['identifier'],
-                            dataset.get('isPartOf')), harvest_job)
-                    continue
+            # old code for previous catalog
+            if p.toolkit.check_ckan_version(max_version='2.7.99'):
```
Comment (adborden) [inline ckanext/datajson/harvester_base.py:271]: I wasn't a fan of this code before, but now intermingling the CKAN 2.8 checks make this even harder to reason about.

Instead of doing it like this, let's create a new harvest class DatajsonHarvester23 for the old CKAN 2.3 code and DatajsonHarvester for the new code. Then you only need to do the version check once to figure out which implementation to use and the readability goes way up. Not to mention you can write a very different algorithm without worrying about CKAN 2.3.

File: ckanext/datajson/tests/test_datajsonharvester.py [hunk 11]
```diff
@@ -41,95 +37,470 @@ def setup(cls):
...
+
+        url = 'http://127.0.0.1:%s/collection-1-parent-2-children.data.json' % mock_datajson_source.PORT
+        obj_ids = self.run_gather(url=url)
+
+        if p.toolkit.check_ckan_version(max_version='2.7.99'):
+            # at CKAN 2.3 with GSA ckanext-harvest fork we expect just parents
+            # after "parents_run" a new job will be raised for children
+            assert_equal(len(obj_ids), 1)
+        else:
+            assert_equal(len(obj_ids), 3)
```
Comment (adborden) [inline ckanext/datajson/tests/test_datajsonharvester.py:168]: oof, yeah, the version check makes the tests very hard to understand. Using multiple class implementations means you can write separate tests for each class. The tests become much more readable.

Since you've already written assertions for 2.3, it would be a nice to keep them :) but not necessary for this work :)

**Comment by avdata99** [comment]
> Thanks @avdata99 , we're getting closer :)
> 
> I'm finding the version checks hard to follow. I understand we're trying to support multiple versions, but when trying to make big changes like this, I think we want to approach this using independent, version-specific implementations for the harvester class.
> 
> Second, I'm having trouble finding the test case for concurrency (e.g. test that harvest import on a child package before a parent throws an error). Can you point it out for me?

This test is still not ready. I read related code and maybe we will need to check if raise the error in this stage. After finish the current failing tests I'll move to this pending test.
About the versioning you are right. It's a good idea. I'll divide this into two different harvesters, one for each version

File: ckanext/datajson/harvester_base.py (hunk 10 above)
Comment (avdata99) [reply ckanext/datajson/harvester_base.py:263]: I cleaned almost all the `check_ckan_version` calls in this file. This one now is the **only** one we use.
Do you think we still need this `DatajsonHarvester23` harvester @adborden?

Comment (adborden) [reply ckanext/datajson/harvester_base.py:263]: Yes, we should still split the implementation at the class level because the tests are still very hard to reason about.

File: bin/travis-build.bash [hunk 12]
```diff
@@ -78,9 +78,23 @@ paster db init -c test-core.ini
 cd ..
 echo "-----------------------------------------------------------------"
 echo "Installing Harvester"
-git clone https://github.com/ckan/ckanext-harvest
+
+git clone https://github.com/GSA/ckanext-harvest
```
Comment (adborden) [inline bin/travis-build.bash:82]: Shouldn't this be ckan/ckanext-harvest for 2.8 and GSA/ckanext-harvest for 2.3?

Comment (avdata99) [reply bin/travis-build.bash:82]: @adborden We are waiting for [a PR](https://github.com/GSA/ckanext-harvest/pull/10) to be merged at upstream.
This is related to our extra notifications.
Until that happens we have our fork using the branch `datagov-catalog`.
